kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/useDevices.js

```diff
@@ -129,51 +143,83 @@ function useDevicesWithFilter(apiParams, filterFunction) {
+function useAsyncDeviceFilter(filterFunction) {
+  const memoized = useMemoize(filterFunction, { getKey: device => device.id });
 
-  return useDevicesWithFilter({}, filterDevices);
+  return async function deviceFilter(device) {
+    try {
+      return await memoized(device);
+    } catch (e) {
+      // If not 404 clear cache to try again on next poll
+      if (e?.response?.status !== 404) {
+        memoized.cache.delete(device.id);
+        throw e;
+      }
+
+      return false;
+    }
+  };
 }
 
 /**
- * Filters devices to those that also have the specified facility
- * @param facilityId
- * @return {{
- *  devices: Ref<NetworkLocation[]>, hasFetched: Ref<bool>, isFetching: Ref<bool>,
- *  filterFailed: Ref<bool>, forceFetch: (function(): Promise<void>)
- * }}
+ * Produces a function that resolves with a boolean for a device that has the specified facility
+ * @param {string|null} [id]
+ * @param {bool|null} [learner_can_sign_up]
+ * @return {function(NetworkLocation): Promise<boolean>}
+ */
+export function useDeviceFacilityFilter({ id = null, learner_can_sign_up = null }) {
+  const filters = {};
+
+  // If `id` is an empty string, we don't want to filter by that
+  if (id) {
+    filters.id = id;
+  }
+
+  if (learner_can_sign_up !== null) {
+    filters.learner_can_sign_up = learner_can_sign_up;
+  }
+
+  if (Object.keys(filters).length === 0) {
+    return () => Promise.resolve(true);
+  }
+
+  return useAsyncDeviceFilter(function deviceFacilityFilter(device) {
+    return deviceHasMatchingFacility(device, filters);
+  });
+}
+
+/**
+ * Produces a function that resolves with a boolean for a device that has the specified channel
+ * @param {string|null} id
+ * @return {function(NetworkLocation): Promise<boolean>}
  */
-export function useDevicesWithFacility({ facilityId = '', soud = false } = {}) {
-  // If facilityId is not provided, then we are at the initial Facility Import workflow
-  // disable any devices unless it is unavailable/offline
-  const filterDevices =
-    facilityId === ''
-      ? () => Promise.resolve(true)
-      : facilityIsAvailableAtDevice.bind({}, facilityId);
-
-  return useDevicesWithFilter({ subset_of_users_device: soud }, filterDevices);
+export function useDeviceChannelFilter({ id = null }) {
+  if (!id) {
+    return () => Promise.resolve(true);
+  }
+
+  return useAsyncDeviceFilter(function deviceChannelFilter(device) {
+    return channelIsAvailableAtDevice(id, device);
+  });
 }
 
 /**
- * Filters devices to those that are not SoUDs/LOD, since SoUD/LOD sync to full-facility devices
- * @return {{
- *  devices: Ref<NetworkLocation[]>, hasFetched: Ref<bool>, isFetching: Ref<bool>,
- *  filterFailed: Ref<bool>, forceFetch: (function(): Promise<void>)
- * }}
+ * Produces a function that resolves with a boolean if Kolibri version is at least the specified
+ * @param {number} major
+ * @param {number} minor
+ * @param {number} patch
+ * @return {function(NetworkLocation): Promise<boolean>}
```

Comment:
This is a beautiful improvement <3

Reply:
I wish it showed the full diff so it was obvious here what it looked like before lol this is really so much nicer a way to do what we need to do here.

Reply:
Yeah by making the filters (filter functions) more like object entities, it disconnects the 'inheritance' structure that was in use before and makes it more flexible